File: scripts/migrate.js
// Minimal forward-only migration runner. Applies every migrations/*.sql that
// hasn't been recorded in the schema_migrations table, in filename order, each
// in its own transaction. No down migrations by design (§ architecture doc).
import { readdir, readFile } from 'node:fs/promises';
import { fileURLToPath } from 'node:url';
import path from 'node:path';
import pg from 'pg';

const { Pool } = pg;
const __dirname = path.dirname(fileURLToPath(import.meta.url));
const migrationsDir = path.join(__dirname, '..', 'migrations');

const databaseUrl = process.env.DATABASE_URL;
if (!databaseUrl) {
  console.error('DATABASE_URL is not set (source your .env or pass it inline)');
  process.exit(1);
}

const pool = new Pool({ connectionString: databaseUrl });

async function run() {
  await pool.query(`
    CREATE TABLE IF NOT EXISTS schema_migrations (
      name TEXT PRIMARY KEY,
      applied_at TIMESTAMPTZ DEFAULT now()
    )
  `);

  const applied = new Set(
    (await pool.query('SELECT name FROM schema_migrations')).rows.map((r) => r.name),
  );

  const files = (await readdir(migrationsDir))
    .filter((f) => f.endsWith('.sql'))
    .sort();

  let count = 0;
  for (const file of files) {
    if (applied.has(file)) continue;
    const sql = await readFile(path.join(migrationsDir, file), 'utf8');
    const client = await pool.connect();
    try {
      await client.query('BEGIN');
      await client.query(sql);
      await client.query('INSERT INTO schema_migrations (name) VALUES ($1)', [file]);
      await client.query('COMMIT');
      console.log(`applied ${file}`);
      count += 1;
    } catch (err) {
      await client.query('ROLLBACK');
      console.error(`failed on ${file}:`, err.message);
      process.exit(1);
    } finally {
      client.release();
    }
  }

  console.log(count === 0 ? 'already up to date' : `applied ${count} migration(s)`);
  await pool.end();
}

run();

File: backend/scripts/migrate.py
"""Minimal forward-only migration runner. Applies every migrations/*.sql not
yet recorded in schema_migrations, in filename order, each in its own
transaction. No down migrations by design.

Run:  python scripts/migrate.py   (from the backend/ dir, with DATABASE_URL set)
"""

from __future__ import annotations

import asyncio
import os
import sys
from pathlib import Path

import asyncpg

# migrations/ lives at the repo root, one level above backend/
MIGRATIONS_DIR = Path(__file__).resolve().parents[2] / "migrations"


async def run() -> None:
    database_url = os.environ.get("DATABASE_URL")
    if not database_url:
        print("DATABASE_URL is not set", file=sys.stderr)
        sys.exit(1)

    conn = await asyncpg.connect(database_url)
    try:
        await conn.execute(
            """
            CREATE TABLE IF NOT EXISTS schema_migrations (
                name TEXT PRIMARY KEY,
                applied_at TIMESTAMPTZ DEFAULT now()
            )
            """
        )
        applied = {r["name"] for r in await conn.fetch("SELECT name FROM schema_migrations")}
        files = sorted(p for p in MIGRATIONS_DIR.glob("*.sql"))

        count = 0
        for path in files:
            if path.name in applied:
                continue
            sql = path.read_text()
            async with conn.transaction():
                await conn.execute(sql)
                await conn.execute(
                    "INSERT INTO schema_migrations (name) VALUES ($1)", path.name
                )
            print(f"applied {path.name}")
            count += 1

        print("already up to date" if count == 0 else f"applied {count} migration(s)")
    finally:
        await conn.close()


if __name__ == "__main__":
    asyncio.run(run())
